bluez-async/src/lib.rs

```diff
@@ -611,43 +673,94 @@ impl BluetoothSession {
         let characteristic = self.characteristic(id);
-        Ok(characteristic.read_value(HashMap::new()).await?)
+        Ok(characteristic
+            .read_value(options_for_offset(offset))
+            .await?)
     }
 
-    /// Write the given value to the given GATT characteristic.
+    /// Write the given value to the given GATT characteristic, with default options.
+    ///
+    /// This is equivalent to calling `write_characteristic_value_with_options(WriteOptions::default())`.
     pub async fn write_characteristic_value(
         &self,
         id: &CharacteristicId,
         value: impl Into<Vec<u8>>,
+    ) -> Result<(), BluetoothError> {
+        self.write_characteristic_value_with_options(id, value, WriteOptions::default())
+            .await
+    }
+
+    /// Write the given value to the given GATT characteristic, with the given options.
+    pub async fn write_characteristic_value_with_options(
+        &self,
+        id: &CharacteristicId,
+        value: impl Into<Vec<u8>>,
+        options: WriteOptions,
     ) -> Result<(), BluetoothError> {
         let characteristic = self.characteristic(id);
         Ok(characteristic
-            .write_value(value.into(), HashMap::new())
+            .write_value(value.into(), options.into())
             .await?)
     }
 
     /// Read the value of the given GATT descriptor.
+    ///
+    /// This is equivalent to calling `read_descriptor_value_with_offset(0)`.
     pub async fn read_descriptor_value(
         &self,
         id: &DescriptorId,
+    ) -> Result<Vec<u8>, BluetoothError> {
+        self.read_descriptor_value_with_offset(id, 0).await
+    }
+
+    /// Read the value of the given GATT descriptor, starting from the given offset.
+    pub async fn read_descriptor_value_with_offset(
+        &self,
+        id: &DescriptorId,
+        offset: usize,
     ) -> Result<Vec<u8>, BluetoothError> {
         let descriptor = self.descriptor(id);
-        Ok(descriptor.read_value(HashMap::new()).await?)
+        Ok(descriptor.read_value(options_for_offset(offset)).await?)
     }
 
     /// Write the given value to the given GATT descriptor.
+    ///
+    /// This is equivalent to calling `write_descriptor_value_with_offset(0)`.
     pub async fn write_descriptor_value(
         &self,
         id: &DescriptorId,
         value: impl Into<Vec<u8>>,
+    ) -> Result<(), BluetoothError> {
+        self.write_descriptor_value_with_offset(id, value, 0).await
+    }
+
+    /// Write the given value to the given GATT descriptor, starting from the given offset.
+    pub async fn write_descriptor_value_with_offset(
+        &self,
+        id: &DescriptorId,
+        value: impl Into<Vec<u8>>,
+        offset: usize,
     ) -> Result<(), BluetoothError> {
         let descriptor = self.descriptor(id);
-        Ok(descriptor.write_value(value.into(), HashMap::new()).await?)
+        Ok(descriptor
+            .write_value(value.into(), options_for_offset(offset))
```

Comment:
I guess .write_value() is generated code?

Reply:
Yes.